Cache and restore solutions on XFEM-affected nodes/elems

bwspenc opened the pull request:
ref #8763

bwspenc committed:
Cache and restore solutions on XFEM-affected nodes/elems ref #8763

We previously stored the dofs to copy the solution from, but sometimes
the dofs changed for existing nodes connected to XFEM elements.  This
caches the nl and aux solutions for all nodes and elements affected
by XFEM, which corrects that issue.  We also get the aux solution on
the new nodes/elems, which we didn't have before.

permcody commented on framework/include/base/XFEMInterface.h:
```diff
@@ -54,6 +55,8 @@ class XFEMInterface : public ConsoleStreamInterface
     _material_data(NULL),
     _bnd_material_data(NULL),
+    _moose_mesh(NULL),
+    _moose_mesh2(NULL),
```
`_moose_mesh2`? This isn't a very descriptive variable name. Why are there two? Can you name it something better?

Unrelated, we prefer `nullptr` over `NULL` in C++11 because of places where we infer type. NULL is actually an `int`, feel free to replace `NULL` when you see it, and avoid adding new ones.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -280,6 +268,26 @@ Node * XFEM::getNodeFromUniqueID(unique_id_type uid)
   return matching_node;
 }
 
+Elem * XFEM::getElemFromUniqueID(unique_id_type uid)
```
Ugh, this probably should be moved into libMesh at some point.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -900,8 +910,24 @@ XFEM::cutMeshWithEFA()
-  //Add new nodes
   const std::vector<EFANode*> NewNodes = _efa_mesh.getNewNodes();
+  const std::vector<EFAElement*> NewElements = _efa_mesh.getChildElements();
+  const std::vector<EFAElement*> DeleteElements = _efa_mesh.getParentElements();
```
What's with the variable names here? They don't follow code standards.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -967,6 +989,31 @@ XFEM::cutMeshWithEFA()
       libmesh_elem->set_node(j) = libmesh_node;
 
+      // Store solution for all nodes affected by XFEM (even existing nodes)
+      if (parent_elem->is_semilocal(_mesh->processor_id()))
```
Wait, are you sure about this? Your processor may not own some of the nodes of a semilocal element. Sure the solution will be correct but do you attempt to write these values later?

bwspenc replied:
The semilocal element might be attached to local nodes.  We err on the side of storing more data than we actually need here, and then only do the setting of the solution on entities that are locally owned.  This is necessary for this to work in cases where the newly-created nodes end up on different processors.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -967,6 +989,31 @@ XFEM::cutMeshWithEFA()
+        {
+          if (nn->id() == libmesh_node->id())
+          {
+            solution_node = _mesh->node_ptr(nn->parent()->id());
```
Just so it's on your radar. This line of code will fail badly if we attempt to run with DistributedMesh. I know that's not your goal here but at least get in the habit of using mooseAssert so that when we try we don't have to dig in with the debugger to find all these problems.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -967,6 +989,31 @@ XFEM::cutMeshWithEFA()
+          }
+        }
+
+        if ((_moose_mesh->isSemiLocal(solution_node)) ||
```
Hmm... This may not be correct. isSemiLocal just means that the element is attached to a local element. libMesh has better methods for testing for ownership, ghosting, evaluable, these days.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
 }
+
+void
+XFEM::saveSolutionForNode(const Node * node_to_store_to,
```
There is a lot of code duplication in these next four methods. Have you considered breaking this into a few more methods and dropping all of the duplicate code?

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
+{
+  const std::vector<MooseVariable *> & vars = sys.getVariables(0);
+  _stored_solution_scratch.clear();
+  _stored_solution_scratch.reserve(2 * vars.size());
```
This is pretty special logic here and isn't documented. It looks like you are doubling storage so you can store current and old values. However you don't say the order in which you store this information. Also, why not store older values? Will that ever be necessary?

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
+    _stored_solution_scratch.push_back(current_solution(dof));
+
+  for (auto dof : _stored_solution_dofs_scratch)
+    _stored_solution_scratch.push_back(old_solution(dof));
```
Alternative self documenting data structure below. I'm fine with either but you should at least mention what you are doing.

```
struct DofValues
{
  Real current;
  Real old;
  Real older;
}

std::map<unique_id_type, DofValues> data;
```

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
+                          const NumericVector<Number> & old_solution)
+{
+  const std::vector<MooseVariable *> & vars = sys.getVariables(0);
+  _stored_solution_scratch.clear();
```
So why is the scratch vector a class member instead of a local member? You are reallocating and copying it around anyway. In fact, you would be better off with using `::assign` here to avoid reallocation here if that's the optimization you are after.

permcody commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
+    for (unsigned int i = 0; i < _stored_solution_dofs_scratch.size(); ++i)
+    {
+      current_solution.set(_stored_solution_dofs_scratch[i], node_solution[i]);
+      old_solution.set(_stored_solution_dofs_scratch[i], node_solution[_stored_solution_dofs_scratch.size() + i]);
```
Yeah, I see it now. There's a little voodoo on this line with your storage indexing. It probably should be documented.

permcody reviewed (commented):
I had several comments. I know this is a bug fix but it seems like there's a lot of quickly developed code here which could be cleaned up and made more general, faster, etc. I'll let you look it over. I don't see anything that prevents merging but there are some maintenance concerns with this new code.

bwspenc commented:
@permcody All good points that I'll address.  This does need some more cleanup -- I wanted some initial feedback on the concept here (and it looks like the tests didn't even pass in my first attempt -- oops!)

I'll summarize the problem: We only add new nodes with XFEM, and they always copy their state from an existing nodes. I previously stored the old node to copy the solution from, but I found that in some cases, the old nodes also ended up with new DOF indices, so their solution values were lost.  It seems like the most reliable thing to do is just cache the solution and then restore it, which ended up being a lot more work than I initially anticipated.  What do you think -- Is that a crazy thing to do? It works reliably now, even on some parallel cases where the crack is running along right next to a processor boundary where it failed before.

Another couple of questions: It looks like the variables can also have components.  Do we ever use more than 1 component in MOOSE?  I'm only caching/recovering the 1st component.

Also, how about the older solutions.  I'm caching/recovering the current and old solution for nl and aux.  Do both of these actually have an older solution?  It looks like aux does, but I'm not sure about nl.  We never use them, so I don't really care about them, but I should probably copy those as well if they're there just so this doesn't unexpectedly break something.

permcody commented:
We do use more than one component in MOOSE for other shape function families: Higher order monomials, hierarchical?, maybe others?

Older solutions do exist for both aux and nl. They are populated and used anytime somebody requests access to those older solutions on either system respectively. 

The reason I picked on you is that I've already been down this cache and restore path for my work. Made all the normal mistakes, fixed many, but there are still a few assumptions, etc.:
See my version here:
https://github.com/idaholab/moose/blob/devel/modules/phase_field/src/postprocessors/GrainTracker.C#L1358-L1436

I also only grab the first component in my implementation as you can see. Mine isn't even as general as yours in that I didn't bother to implement an elemental version. Also, looking at my code, I recall that setting an old or older solution on a steady state solve is bad (like seg fault bad). 

All of this caching code really should be put in the framework eventually so we don't end up writing it yet again for some other task.

bwspenc commented:
@permcody Thanks for the info. If you've done this another place, I guess I'm not crazy.  It's going to be next week before I can get back to this, but I'm happy to clean this up then before you merge this.  @jiangwen84 you can just grab this commit to use for the milestone work in the mean-time.

bwspenc committed:
Address review comments and generalize ref #8763

bwspenc committed:
Add tests for XFEM solution initialization closes #8763

bwspenc commented:
@permcody I've significantly re-worked this and I believe that I've addressed your comments.  I added a test that confirms that this works for both nodal and element variables, including ones with multiple components.  Without this fix, this test fails in parallel in multiple ways.

permcody reviewed (approved):
This looks better. At some point in the future we'll still have to factor out all this solution update logic though. Not today...
